Frameworks/CoreGraphics/DWriteFontBinaryDataCollectionLoader.mm

```diff
@@ -0,0 +1,170 @@
+
+using namespace Microsoft::WRL;
+
+/**
+ * Private helpers to append a CFErrorRef or nullptr to the end of a CFMutableArray if it exists
+ */
+static inline void __AppendErrorIfExists(CFMutableArrayRef errors, CFIndex errorCode) {
+    if (errors) {
+        woc::unique_cf<CFErrorRef> error{ CFErrorCreate(nullptr, kCFErrorDomainCocoa, errorCode, nullptr) };
+        CFArrayAppendValue(errors, error.get());
+    }
+}
+
+static inline void __AppendNullptrIfExists(CFMutableArrayRef errors) {
+    if (errors) {
+        CFArrayAppendValue(errors, nullptr);
+    }
+}
+
+/**
+ * Private class.
+ * Custom implementation of IDWriteFontFileEnumerator, which enumerates over internal data and returns a font file
+ */
+class DWriteFontFileEnumerator : public RuntimeClass<RuntimeClassFlags<WinRtClassicComMix>, IDWriteFontFileEnumerator> {
+protected:
+    InspectableClass(L"Windows.Bridge.DirectWrite", TrustLevel::BaseTrust);
+
+public:
+    DWriteFontFileEnumerator() {
+    }
+
+    HRESULT RuntimeClassInitialize(CFArrayRef fontDatas, std::vector<ComPtr<IDWriteFontFile>>* previouslyCreatedFiles) {
+        CFRetain(fontDatas);
+        m_fontDatas.reset(fontDatas);
+        m_previouslyCreatedFiles = previouslyCreatedFiles;
+        return S_OK;
+    }
+
+    HRESULT STDMETHODCALLTYPE GetCurrentFontFile(_Out_ IDWriteFontFile** fontFile) {
+        RETURN_HR_IF_NULL(E_POINTER, fontFile);
+        RETURN_HR_IF(E_ILLEGAL_METHOD_CALL, m_location < 0 || m_location > CFArrayGetCount(m_fontDatas.get()));
+
+        if (0 <= m_location && m_location < m_previouslyCreatedFiles->size()) {
+            m_previouslyCreatedFiles->at(m_location).CopyTo(fontFile);
+        } else {
+            ComPtr<IDWriteFactory> dwriteFactory;
+            RETURN_IF_FAILED(DWriteCreateFactory(DWRITE_FACTORY_TYPE_SHARED, __uuidof(IDWriteFactory), &dwriteFactory));
+
+            ComPtr<DWriteFontBinaryDataLoader> loader;
+            RETURN_IF_FAILED(MakeAndInitialize<DWriteFontBinaryDataLoader>(&loader,
+                                                                           static_cast<CFDataRef>(
+                                                                               CFArrayGetValueAtIndex(m_fontDatas.get(), m_location))));
+            RETURN_IF_FAILED(dwriteFactory->RegisterFontFileLoader(loader.Get()));
+
+            int unused;
+            RETURN_IF_FAILED(dwriteFactory->CreateCustomFontFileReference(&unused, sizeof(unused), loader.Get(), fontFile));
+            m_previouslyCreatedFiles->emplace_back(*fontFile);
+        }
+
+        return S_OK;
+    }
+
+    HRESULT STDMETHODCALLTYPE MoveNext(_Out_ BOOL* hasCurrentFile) {
+        RETURN_HR_IF_NULL(E_POINTER, hasCurrentFile);
+        *hasCurrentFile = ++m_location < CFArrayGetCount(m_fontDatas.get());
+        return S_OK;
+    }
+
+private:
+    int m_location = -1;
+
+    woc::unique_cf<CFArrayRef> m_fontDatas;
+    std::vector<ComPtr<IDWriteFontFile>>* m_previouslyCreatedFiles;
```

Comment:
"shouldn't be owning at all" ... the class blindly uses the pointer which implies something about the lifetime of that memory. This is pretty much the definition of shared ownership. It is saying, "I didn't create the thing, but I have a reasonable expectation that whenever I want, I can use it and therefore store it among my *member* variables, the set of objects I own and manage". To correctly make that work, you need something that enforces that. Right now its a tenuous contract with the singleton parent object that requires a future code maintainer to read  and understand all of the dwrite loader classes and uses as there is nothing else linking the two. The "child" class is not a child in any traditional sense. Its lifetime is not tied to the parent, its definition is not private to the parent, etc. 

This makes the contract hard to continue in the future as someone may invalidate the baseline assumption and then be real confused. One of the *amazing* things about shared_ptr and friends is that the type itself implies the ownership so its much harder to do the wrong thing and at all times the use is clear and communicated. 

My 2 cents is that code should be correct first, maintainable second, and fast/clever/whatever third. If you think that the shared_ptr is overly cumbersome, please take performance measurements to show its impact and we can evaluate from there. Until then, lets make it as easy as possible for the next person.

---
In reply to: [90944931](https://github.com/Microsoft/WinObjC/pull/1504#discussion_r90944931) [](ancestors = 90944931,90943752,90943527,90917725)